### PR TITLE
Move from kotlinx.dom to Jackson for pom parsing

### LINKED ISSUE #4: Parse external JavaDoc links from poms
Will allow generated docs to link to adventure and etc.

### LINKED ISSUE #6: Parse and replace maven project properties in poms

### PR DESCRIPTION
**Pull request by jpenilla**
Closes #4

### CODE REVIEW

File: src/main/kotlin/xyz/jpenilla/toothpick/ConfigureSubprojects.kt [hunk 1]
```diff
@@ -120,41 +118,14 @@ private fun Project.configureServerProject() {
...
-                shadedPattern.replace("\${minecraft_version}", toothpick.nmsPackage),
-                rawString,
-                excludes = excludes
-              )
-            )
-          }
-        }
+    // Import relocations from server pom
+    val pom = subproject.pom ?: return@getting
+    val shadePlugin: ShadePlugin = pom.build.plugins.find { it is ShadePlugin } as ShadePlugin? ?: error("Could not find shade plugin in server pom!")
```
Comment (Proximyst) [inline src/main/kotlin/xyz/jpenilla/toothpick/ConfigureSubprojects.kt]:
```suggestion
    val shadePlugin: ShadePlugin = pom.build.plugins.filterIsInstance().firstOrNull() ?: error("Could not find shade plugin in server pom!")
```

File: src/main/kotlin/xyz/jpenilla/toothpick/data/Pom.kt [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+internal data class DependencyManagement(
+  val dependencies: List<Dependency> = emptyList()
+)
+
+internal data class Repository(
+  val id: String,
+  val url: String
+)
+
+internal data class Dependency(
```
Comment (Proximyst) [inline src/main/kotlin/xyz/jpenilla/toothpick/data/Pom.kt:51]: This does not fulfill the entirety of the definition of dependencies. Consider ignoring unknown properties or filling it: https://maven.apache.org/pom.html#Dependencies

File: src/main/kotlin/xyz/jpenilla/toothpick/data/Pom.kt [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  val dependencies: List<Dependency> = emptyList(),
+  val dependencyManagement: DependencyManagement?,
+  val build: BuildSection
+)
+
+internal data class DependencyManagement(
+  val dependencies: List<Dependency> = emptyList()
+)
+
+internal data class Repository(
```
Comment (Proximyst) [inline src/main/kotlin/xyz/jpenilla/toothpick/data/Pom.kt:45]: This does not fulfill the entirety of the definition of repositories. Consider ignoring unknown properties or filling it: https://maven.apache.org/pom.html#Repositories

File: src/main/kotlin/xyz/jpenilla/toothpick/data/ShadePlugin.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package xyz.jpenilla.toothpick.data
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties
+
+internal data class ShadePlugin(
```
Comment (Proximyst) [inline src/main/kotlin/xyz/jpenilla/toothpick/data/ShadePlugin.kt:29]: This does not fulfill the entirety of the shade plugin parameters. Consider ignoring unknown parameters or filling it: https://github.com/apache/maven-shade-plugin/blob/master/src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java#L99-L375

File: src/main/kotlin/xyz/jpenilla/toothpick/data/ShadePlugin.kt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  val filters: List<Filter> = emptyList()
+)
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+internal data class ShadeExecution(
+  val phase: String,
+  val configuration: ShadeConfiguration = ShadeConfiguration()
+)
+
+internal data class Filter(
```
Comment (Proximyst) [inline src/main/kotlin/xyz/jpenilla/toothpick/data/ShadePlugin.kt:51]: This does not fulfill the entirety of the filter parameters. Consider ignoring unknown properties or filling it: https://github.com/apache/maven-shade-plugin/blob/master/src/main/java/org/apache/maven/plugins/shade/mojo/ArchiveFilter.java#L29-L35

**Review by Proximyst — CHANGES_REQUESTED** [review]
Looks good :)

File: src/main/kotlin/xyz/jpenilla/toothpick/data/ShadePlugin.kt (hunk 4 above)
Comment (jpenilla) [reply src/main/kotlin/xyz/jpenilla/toothpick/data/ShadePlugin.kt:29]: Seems like it inherits `@JsonIgnoreProperties(ignoreUnknown = true)` from it's superclass

Comment (Proximyst) [reply src/main/kotlin/xyz/jpenilla/toothpick/data/ShadePlugin.kt:29]: Jackson does not inherit annotations, sadly. It must be explicitly defined.

**Comment by kennytv** [comment]
thank you jomp, very cool

**Pull request merged**